### ISSUE TITLE
opensource 를 활용한 LLM model 학습 및 준비 과정

### ISSUE THREAD

**Issue by jms0522** [jms0522/Medical-project#39]
- open source LLM 활용하여 의료 도메인을 학습 시켜 모델 비교와 결과를 도출하기

참고 : #40

**Comment by jms0522** [comment]
LLAMA2 기반의 ko-llama2 사용해봤지만 성능 낮음

**Comment by jms0522** [comment]
# 0218 LDCC-SOLAR-10.7B
- ko LLM dash board 1위  https://huggingface.co/LDCC/LDCC-SOLAR-10.7B
- [yanolja/KoSOLAR-10.7B-v0.1](https://huggingface.co/yanolja/KoSOLAR-10.7B-v0.1) 이 모델에 finetune (현재 이 모델은 사용 불가)
- 테스트 용으로 동일하게  royboy0416/ko-alpaca 데이터 셋을 추가 학습

**Comment by jms0522** [comment]
# 0220 

- api gemini를 사용했으나 성능보다 구현이 목적이고 좀 더 실무에 맞을 거 같아 모델 성능은 좀 포기하고 오픈소스로 제작 결정.
- **모델보다 자동화 파이프라인에 집중하고 여러가지 생성 데이터와 외부데이터를 수집, 전처리 , 벡터 store 저장하는 과정에 집중하기로 함.**

**Comment by jms0522** [comment]
# 0221 processing …

지식인 QnA 데이터와 네이버 질병 백과 질병 사전 목록을 통한 LLM 모델 fine turn을 위한 데이터 전처리 작업 진행.

답변에서 병명 추출 / 데이터 밸런싱 / 한국어와 띄어쓰기 (딥러닝 모델) 교정 /
한 질문에 다양한 병명이 있을 경우 score를 줘서 가장 높은 score의 병명을 추출 / 특수 문자 제거 / 증상에 대한 score가 너무 낮은 데이터 삭제 / 증상과 정의 분리 / 증상 컬럼 없는 경우 정의 컬럼으로 대체 / 증상 정의 둘 다 없는 경우 삭제 / 소아기칭, 영유아복통 등 세분화된 질병 삭제 / word count를 이용한 밸런싱 등 
답 없는 질문과 답변 내용을 최대한 전처리 하기 위해 노력.


전체 데이터 : 약 300만 (지식인 QnA 한정)
시험 전처리 : 100만개 -> 6만개로 전처리

앞으로의 과제 : 답변 수가 너무 적은 데이터를 삭제 하는 방법보다
데이터 증대 / 질문의 유사한 유의어, 혹은 단어 삽입, 위치 변경, 랜덤 단어 삭제 등으로 문맥이 비숫한 질문을 만들어 증강을 생각 해 볼 수 있고, bert gtp2등의 모델을 fine tune 하여 문맥이 유사한 질문을 생성으로 증강 고려 할 수 있음.

느낀점 : 데이터가 너무 정형화되지 않고 증상을 물어보는 질문도 너무 부정확해서 전처리 하는 게 너무 힘들다.
사실 전처리를 통해 이만큼까지 데이터를 줄이는 것도 데이터를 훼손 시키고 변형으로 인한 손실이 크지만 현재 kict deep learning server에서도 memory 이슈가 계속 발생하는 상황.
solar, llama2 등등.
peft LoRa의 방법론을 더 깊게 파서 경량화를 더 해보든
애초에 토큰의 수가 적은 모델을 사용하던
돈을 발라서 학습을 시키던 선택을 해야할 듯.

사실 모델을 돌려서 평가를 해봐야 전처리든 미세조정이든
할텐데 조금 답답한 상황이다.
결국 또 RAG가 답 일지도..
그럼 질문과 답변을 전처리해서 vectort store에 넣어야 하는지 아직 확신이 안 듬..

**Comment by jms0522** [comment]
#0222 solar -10.7b training
<img width="1196" alt="solar_training0222" src="https://github.com/jms0522/Medical-project/assets/132258503/18dc8f15-68bf-4569-acac-77c751f60aa3">

- 훈련 중인데 언제 끝날지 모르겠음. 리소스가 부족하지 않을지도 모르겠어서 끝나면 결과 공유함.

**Comment by minseok1994** [comment]
https://kakaoenterprise.github.io/deepdive/210525#b6b2:fn-back:14
<img width="923" alt="image" src="https://github.com/jms0522/Medical-project/assets/145941741/adc5ba93-3ffe-44cc-b15b-077edd1b5706">
- 카카오 엔터프라이즈에서 사용하는 8가지 자동화 평가지표

https://huggingface.co/spaces/evaluate-metric/bertscore
<img width="1081" alt="image" src="https://github.com/jms0522/Medical-project/assets/145941741/9228abd5-331f-4dfc-9fa9-085a8bf977b5">
- 그 중 BERT Score만 가져와서 직접 실험

**Comment by stfano** [comment]
# autotrain-advanced를 이용한 finetuning 진행 

## 파라미터 및 소요시간 정보
- peft기법
- lora_alpha = 16
- batch_size = 1
- epochs = 1
- model : squarelike/llama2-ko-medical-7b
- google colab t4 2시간 30분 소요
- 
<img width="653" alt="스크린샷 2024-02-22 오후 6 09 35" src="https://github.com/jms0522/Medical-project/assets/57385696/2fc3784e-c60a-470d-95f0-4caf08f1dd67">

## bert_score 
<img width="692" alt="스크린샷 2024-02-22 오후 5 44 17" src="https://github.com/jms0522/Medical-project/assets/57385696/ccd560a4-3cc0-4d5b-81ad-136c66755d9e">

**Comment by jms0522** [comment]
## 0222 information

- 성능 향상을 위한 이론 : [https://www.dbpia.co.kr/Journal/articleDetail?nodeId=NODE11043897](url)

**Comment by jms0522** [comment]
## google 신규 LLM gemma

- open source
- 뛰어난 성능 
- 작은 모델 2b 존재

![image](https://github.com/jms0522/Medical-project/assets/132258503/04e92655-37f4-401e-9916-ba7bb03c139f)

![image](https://github.com/jms0522/Medical-project/assets/132258503/569f4a00-b1a3-4a58-8a34-0626e4a1548a)

![image](https://github.com/jms0522/Medical-project/assets/132258503/a8fa3c13-f5ec-43b3-b582-5cee91a0be4f)

- 아직 한국어 fine tune 모델의 부재


![image](https://github.com/jms0522/Medical-project/assets/132258503/494dbfeb-4e90-4e1c-a3e2-d5caf61f9cc4)

![image](https://github.com/jms0522/Medical-project/assets/132258503/8de81c6f-ad47-4891-b2c5-b79d43b4f1bb)

**Comment by stfano** [comment]
# huggingface에 모델올리는법 

- huggingface에 model repository 만들기
- 로컬에서 git lfs install
- huggingface-cli login 해서 token 입력
- git clone <https://huggingface.co/USER_NAME/YOUR_MODEL_NAME> 생성한 모델 이름 입력
## model repo 클론 및 모든 파일 add 
<img width="639" alt="image" src="https://github.com/jms0522/Medical-project/assets/57385696/39313793-a179-4f6e-9415-3eac226fa30c">

## commit 및 push 진행 
<img width="526" alt="image" src="https://github.com/jms0522/Medical-project/assets/57385696/6de049fa-4bda-4ffd-97bf-01ed248a139c">

## 올라온 모델 확인 
<img width="765" alt="image" src="https://github.com/jms0522/Medical-project/assets/57385696/d6bf3b51-6793-4c81-8f2e-0a18e9596cc3">

**Comment by jms0522** [comment]
## 모델 학습 인프라 구축 (kict 연결)

-대용량 모델을 데이터셋에 학습을 시키려니 많인 리소스가 필요함
- k-ict (빅데이터 센터) 에서 제공하는 딥러닝 분석 서버(GPU : M10#02)을 팀원마다 신청, 할당 받아서 사용
- 한 서버당 모델 학습에 필요한 메모리가 충분치 않아서 4개의 서버를 연결하려 시도 중. (@cheol2Y )

<img width="893" alt="스크린샷 2024-02-25 18 56 54" src="https://github.com/jms0522/Medical-project/assets/132258503/2d94e899-1ca8-47ac-8755-4d7e292eed20">
